pkg/koordlet/qosmanager/plugins/cpuevict/cpu_evict.go

```diff
@@ -0,0 +1,383 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cpuevict
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/component-base/featuregate"
+	"k8s.io/klog/v2"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+	slov1alpha1 "github.com/koordinator-sh/koordinator/apis/slo/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/features"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/common/reason"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/common/utils"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/metriccache"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/qosmanager/k8s"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/qosmanager/metricsquery"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/qosmanager/plugins"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/statesinformer"
+	"github.com/koordinator-sh/koordinator/pkg/util"
+)
+
+const (
+	beCPUSatisfactionLowPercentMax   = 60
+	beCPUSatisfactionUpperPercentMax = 100
+	beCPUUsageHighEnough             = 0.9
+)
+
+const (
+	DefaultReconcileIntervalInSeconds    = 5
+	DefaultCollectMetricsIntervalSeconds = int64(1)
+	DefaultCPUEvictCoolTimeSeconds       = 20
+)
+
+type Config struct {
+	ReconcileIntervalInSeconds    int   `json:"reconcileIntervalInSeconds"`
+	CollectMetricsIntervalSeconds int64 `json:"collectMetricsIntervalSeconds"`
+	CPUEvictCoolTimeSeconds       int   `json:"cpuEvictCoolTimeSeconds"`
+}
+
+type CPUEvictor struct {
+	stopCh chan struct{}
+	cfg    *Config
+
+	k8sClient      k8s.K8sClient
+	statesInformer statesinformer.StatesInformer
+	metricCache    metriccache.MetricCache
+	metricsQuery   metricsquery.MetricsQuery
```

Comment:
k8sClient, statesInformer, metricCache can be replaced by the new `Context`.

how about we first establish the framework of qos-manager first, then mv each plugin one by one, since some plugin should be mv as hook proxy plugins.